### ISSUE TITLE
Memory leak when importing mongodb-memory-server

### ISSUE THREAD

**Issue by exoer** [nodkz/mongodb-memory-server#276]
## Versions

- NodeJS: 13.10.1
- mongodb-memory-server-*: 6.3.2
- mongodb: 3.5.4
- system: MacOS Catalina

package: mongo-memory-server 

## What is the Problem?

Memory leak when importing `mongodb-memory-server` module.
When using jest the heap size keeps growing with each run. 

## Code Example

`module1.js`. This demonstration just imports the memory-server. It doesn't even need to instanciate a server to cause the memory leak
```
import MongoMemoryServer from "mongodb-memory-server"
import mongodb from 'mongodb'

export const example = ()=>{
  return {
    f1: ()=> {return 1},
  }
}
export default example
```

in my test `mod1.spec.js`
```
import ex from './module1' // only import the module
describe('module', ()=>{
  it('cleans up', ()=> {
    let a = new Array(1e1)
    expect(a.length).toEqual(1e1)
  })
})

```
I run the test from `package.json`

```
  "scripts": {
    "test": "NODE_ENV=test jest --watchAll --runInBand --detectOpenHandles --logHeapUsage"
  },

```

i get this output after running a dozen times
```
 PASS  ./mod1.spec.js (443 MB heap size)
 PASS  ./mod2.spec.js (453 MB heap size)
 PASS  ./mod3.spec.js (453 MB heap size)

Test Suites: 3 passed, 3 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        0.475s, estimated 1s
Ran all test suites.

Watch Usage: Press w to show more.
```
The heap size keeps growing each time i run the test suite

After a handful more runs
```
 PASS  ./mod1.spec.js (658 MB heap size)
 PASS  ./mod2.spec.js (668 MB heap size)
 PASS  ./mod3.spec.js (658 MB heap size)

Test Suites: 3 passed, 3 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        0.679s, estimated 1s
Ran all test suites.

Watch Usage: Press w to show more.

```

## Do you know *why* it happenes?

*no*

**Comment by hasezoey** [comment]
cant reproduce (using `mongo-memory-server-core`'s tests)
System: Linux Mint 19.3 (kernel 5.3)
MongoDB: 4.0.3 (default)
NodeJS 12.16.1
Commit: a498bef1ab18d6bc229c76551ebc7ce4289af1b9 (latest currently available)

The heap size for consecutive runs & runInBand is between ~50mb (tests without spinning up any instance) to ~70mb for single-instances to ~100mb for replica sets (at least 3 instances)

The heap size for consecutive runs & no runInBand is between ~45mb (tests without spinning up any instance) to ~65mb for single-instances to ~55mb for replica sets (at least 3 instances)

**Comment by exoer** [comment]
Thank you hasezoey for trying to reproduce.

I've created a small project on github (with a `docker-compose.yml`) to reproduce:
https://github.com/exoer/mongo-memory-server-leak

**Comment by hasezoey** [comment]
@exoer i noticed you dont have any jest config, and so run it in `isolatedModules: false` and use `watch` instead of "consecutive runs"

i ran the tests agains (the project ones) and with `watchAll` and `isolatedModules: true`, i couldnt reproduce it
then i ran it again with `isolatedModules: false` and the results were on the first run all about ~100mb and on re-triggering the run (in watch mode) all were in the 50-60mb range again

---

now to your project:
i noticed that you dont use `ts-jest` and jest directly
i could otherwise observe what you meant, and i debugged it a litte, here are my steps i took:
- removed all dependencies that were not in use (`mongo-mock`, `mock-mongo`, `redis`, `redis-async`)
- removed `mongo-memory-server-core` from the devDependencies because it was already in the normal dependencies
- removed all babel related things and rewrote the files to use the require structure (because import dosnt work without babel / typescript)
- removed mongodb dependency from being imported
- and finally removed `mongo-memory-server` from being called at all

and after all these steps the head still gone up and up

TL;DR: this donst seem like an problem with `mongo-memory-server`

**Comment by exoer** [comment]
@hasezoey  Thank you very much for testing this. 

Those other dependencies (`mongo-mock, mock-mongo, redis, redis-async`) were used for other tests and i forgot to remove them.

You write that removing `mongo-memory-server` also increase the memory. I have observed that too. 

But when I run the tests enough times (10-ish) without importing `mongodb-memory-server-core` it looks like `jest` manages to garbage collect (because the heap size can drop significantly). But if I import the `mongodb-memory-server-core` I have never seen that the memory heap gets garbage collected. 


### Example
Running tests with import
```
import {MongoMemoryServer} from "mongodb-memory-server-core"
```
until the heapsize got to around 450 MB. Then uncommenting the import. The heap size kept increasing to 712 MB. Then it dropped (garbage collected i assume?). 
```
nodeapp_1  | PASS ./mod2.spec.js (671 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (672 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (679 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.306s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (680 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (687 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (688 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.318s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (695 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (696 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (703 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.321s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (704 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (710 MB heap size)
nodeapp_1  | 
nodeapp_1  | PASS ./mod1.spec.js (712 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.319s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | PASS ./mod2.spec.js (463 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (470 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (471 MB heap size)
```
I have tried with importing other dependencies as well, but so far my suspicion is on 'mongodb-memory-server`. But i can't prove it. Any suggestions?

### Example 2
Running test where no dependences were imported. The heap increases slowly, and can drop to 50MB after a handful of runs. 
```
nodeapp_1  | Time:        0.264s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (62 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (64 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (66 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.258s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (52 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (53 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (55 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.251s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (61 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (62 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (68 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.245s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (74 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (66 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (73 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.278s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (79 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (80 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (82 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.257s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (87 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (93 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (95 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.249s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (100 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (101 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (108 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.248s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (49 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (55 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (57 MB heap size)
nodeapp_1  | 
nodeapp_1  | Test Suites: 3 passed, 3 total
nodeapp_1  | Tests:       7 passed, 7 total
nodeapp_1  | Snapshots:   0 total
nodeapp_1  | Time:        0.251s, estimated 1s
nodeapp_1  | Ran all test suites.
nodeapp_1  | 
nodeapp_1  | PASS ./mod2.spec.js (62 MB heap size)
nodeapp_1  | PASS ./mod1.spec.js (64 MB heap size)
nodeapp_1  | PASS ./mod3.spec.js (70 MB heap size)
nodeapp_1  | 
```

I've tried to look into chrome debug console, but it is not easy for me to decipher it. I've seen `graceful-fs` and `BSON` (related to Mongo i suppose?) taking up much memory. 

I am not using typescript so I can't test with `jest-ts` and `isolatedModules`

**Comment by hasezoey** [comment]
@exoer from what i could find is that an common problem is `graceful-fs`, but i dont really understand why it is still happening when no other module is require/imported
and yes `BSON` is an extended JSON used by mongodb

**Comment by exoer** [comment]
I found out that the `jest` option `--detectOpenHandles` was probably the cause of this. When removing this option the heap size stopped growing with each run.

**Comment by hasezoey** [comment]
@exoer i never used `detectOpenHandles` in my test runs, so this is probably not the cause

**Comment by timleslie** [comment]
FWIW I am seeing a similar memory leak that appears to be related to simply importing `mongodb-memory-server-core`. I will try to provide more details if I can isolate the problem further.

**Comment by exoer** [comment]
In retrospect i think that the memoryleak in was because i didn't close the `server` (instance of MongoMemoryServer) after the test finished. I only closed the `connection`. Now when closing the server instance i don't get any memory leaks.

**Comment by hasezoey** [comment]
@exoer yes, this happened in the package's tests many times too, and somehow `detectOpenHandles` just stalled jest (known problem there for an long time) and was annoying to pin down (no clues, had to isolate run tests)
